src/Reflex/Dom/Contrib/Router.hs

```diff
@@ -3,81 +3,191 @@
+import qualified Web.Routes.PathInfo       as WR
 #if ghcjs_HOST_OS
-import           Data.Maybe                (fromJust)
-import           Control.Monad.IO.Class    (liftIO)
 import qualified GHCJS.DOM                 as DOM
 import qualified GHCJS.DOM.Document        as DOM
-import qualified GHCJS.DOM.EventM          as DOM
+import           GHCJS.DOM.EventM          (on)
 import           GHCJS.DOM.History         (back, forward, pushState)
-import           GHCJS.DOM.Location        (toString)
+import           GHCJS.DOM.Location        (getPathname, toString)
 import           GHCJS.DOM.Window          (Window, getHistory,
                                             getLocation, popState)
 import           GHCJS.Marshal.Pure
 #else
+import           Control.Monad.Reader      (ReaderT)
 #endif
 
 ------------------------------------------------------------------------------
-data RouteConfig t = RouteConfig
+data RouteConfig t a = RouteConfig
   { _routeConfig_forward   :: Event t () -- ^ Move the browser history forward
   , _routeConfig_back      :: Event t () -- ^ Move the browser history back
-  , _routeConfig_pushState :: Event t Text -- ^ Push to the URL state
-  -- , _routeConfig_pathBase  :: Text
-  --   -- ^ The part of the URL not related to SPA routing
+  , _routeConfig_pushState :: Event t a  -- ^ Push to the URL state
   }
 
-instance Reflex t => Default (RouteConfig t) where
+makeLenses ''RouteConfig
+
+instance Reflex t => Default (RouteConfig t a) where
   def = RouteConfig never never never
 
-instance Reflex t => Monoid (RouteConfig t) where
+
+instance Reflex t => Monoid (RouteConfig t a) where
   mempty = def
   mappend (RouteConfig f1 b1 p1) (RouteConfig f2 b2 p2) =
     RouteConfig (mappend f1 f2) (mappend b1 b2) (leftmost [p1, p2])
 
-  mconcat rcs =
-    RouteConfig (mconcat $ map _routeConfig_forward rcs)
-                (mconcat $ map _routeConfig_back rcs)
-                (leftmost $ map _routeConfig_pushState rcs)
 
-data Route t = Route {
-    _route_value :: Dynamic t Text -- ^ URL value
+data Route t a = Route {
+    _route_value   :: Dynamic t (Either T.Text a)     -- ^ Routing value
+  , _route_fullURI :: Dynamic t (U.URIRef U.Absolute) -- ^ Full URI
   }
 
-instance HasValue (Route t) where
-  type Value (Route t) = Dynamic t Text
+makeLenses ''Route
+
+instance HasValue (Route t a) where
+  type Value (Route t a) = Dynamic t (Either T.Text a)
   value = _route_value
 
--- | Manipulate and track the URL text for dynamic routing of a widget
-route :: (HasWebView m, MonadWidget t m) => RouteConfig t -> m (Route t)
-#if ghcjs_HOST_OS
-route (RouteConfig goForward goBack sSet) = do
-  win <- askDomWindow
-  loc <- getLocation' win
+-------------------------------------------------------------------------------
+-- | Manipulate and track the URL 'GHCJS.DOM.Types.Location' for dynamic
+--   routing of a widget
+route
+  :: (HasWebView m, MonadWidget t m)
+  => (Location -> IO (Either T.Text a))
```

Comment:
`Location` is basically `ghcjs-dom`'s version of `URIRef Absolute`. It's already parsed into protocol, origin, auth, port, path, fragment for us.